File: backend/api/inbox.py
from datetime import datetime
from typing import Optional
from fastapi import APIRouter, Depends, HTTPException
from sqlmodel import select, func

from api.auth import get_current_user
from database import get_session
from models import InboxMessage, User

router = APIRouter(prefix="/inbox", tags=["inbox"])


def _to_dict(m: InboxMessage) -> dict:
    return {
        "id": m.id,
        "company_id": m.company_id,
        "recipient_user_id": m.recipient_user_id,
        "source_type": m.source_type,
        "source_id": m.source_id,
        "title": m.title,
        "body": m.body,
        "read": m.read,
        "created_at": m.created_at.isoformat(),
    }


@router.get("")
def list_inbox(
    company_id: Optional[str] = None,
    unread_only: bool = False,
    period: Optional[str] = None,   # "week" | "month"
    current_user: User = Depends(get_current_user),
):
    with get_session() as session:
        q = select(InboxMessage).where(InboxMessage.recipient_user_id == current_user.id)
        if company_id:
            q = q.where(InboxMessage.company_id == company_id)
        if unread_only:
            q = q.where(InboxMessage.read == False)
        if period == "week":
            from datetime import timedelta
            q = q.where(InboxMessage.created_at >= datetime.utcnow() - timedelta(days=7))
        elif period == "month":
            from datetime import timedelta
            q = q.where(InboxMessage.created_at >= datetime.utcnow() - timedelta(days=30))
        q = q.order_by(InboxMessage.created_at.desc())
        return [_to_dict(m) for m in session.exec(q).all()]


@router.get("/unread-count")
def unread_count(company_id: Optional[str] = None, current_user: User = Depends(get_current_user)):
    with get_session() as session:
        q = select(func.count()).where(
            InboxMessage.recipient_user_id == current_user.id,
            InboxMessage.read == False,
        )
        if company_id:
            q = q.where(InboxMessage.company_id == company_id)
        count = session.exec(q).one()
        return {"count": count}


@router.post("/{msg_id}/read")
def mark_read(msg_id: str, current_user: User = Depends(get_current_user)):
    with get_session() as session:
        msg = session.get(InboxMessage, msg_id)
        if not msg or msg.recipient_user_id != current_user.id:
            raise HTTPException(status_code=404, detail="Message not found")
        msg.read = True
        session.add(msg)
        session.commit()
        session.refresh(msg)
        return _to_dict(msg)


@router.post("/read-all")
def mark_all_read(company_id: Optional[str] = None, current_user: User = Depends(get_current_user)):
    with get_session() as session:
        q = select(InboxMessage).where(
            InboxMessage.recipient_user_id == current_user.id,
            InboxMessage.read == False,
        )
        if company_id:
            q = q.where(InboxMessage.company_id == company_id)
        msgs = session.exec(q).all()
        for m in msgs:
            m.read = True
            session.add(m)
        session.commit()
        return {"marked": len(msgs)}


@router.delete("/{msg_id}", status_code=204)
def delete_message(msg_id: str, current_user: User = Depends(get_current_user)):
    with get_session() as session:
        msg = session.get(InboxMessage, msg_id)
        if not msg or msg.recipient_user_id != current_user.id:
            raise HTTPException(status_code=404, detail="Message not found")
        session.delete(msg)
        session.commit()

File: frontend/src/lib/api/inbox.ts
import { api } from './client';

export type InboxMessage = {
	id: string;
	company_id: string;
	recipient_user_id: string;
	source_type: 'flow' | 'task_request' | 'task_result' | 'system';
	source_id: string | null;
	title: string;
	body: string;
	read: boolean;
	created_at: string;
};

export type TaskRequest = {
	id: string;
	company_id: string;
	requester_user_id: string;
	department_id: string | null;
	skill_filter: string | null;
	assigned_agent_id: string | null;
	responsible_user_id: string | null;
	title: string;
	body: string;
	human_note: string | null;
	status: 'pending' | 'assigned' | 'running' | 'completed' | 'rejected';
	result: string | null;
	created_at: string;
	updated_at: string;
};

function qs(p: Record<string, string | boolean | undefined>) {
	const u = new URLSearchParams();
	for (const [k, v] of Object.entries(p)) {
		if (v !== undefined && v !== '') u.set(k, String(v));
	}
	return u.toString() ? `?${u}` : '';
}

export const inbox = {
	list:        (params?: { company_id?: string; unread_only?: boolean; period?: string }) =>
		api.get<InboxMessage[]>(`/inbox${qs(params ?? {})}`),
	unreadCount: (company_id?: string) =>
		api.get<{ count: number }>(`/inbox/unread-count${company_id ? `?company_id=${company_id}` : ''}`),
	markRead:    (id: string)         => api.post<InboxMessage>(`/inbox/${id}/read`, {}),
	markAllRead: (company_id?: string) => api.post<{ marked: number }>(`/inbox/read-all${company_id ? `?company_id=${company_id}` : ''}`, {}),
	delete:      (id: string)         => api.delete(`/inbox/${id}`),
};

export const taskRequests = {
	list:   (params?: { company_id?: string; status?: string }) =>
		api.get<TaskRequest[]>(`/task-requests${qs(params ?? {})}`),
	create: (body: { company_id: string; department_id?: string; skill_filter?: string; title: string; body: string }) =>
		api.post<TaskRequest>('/task-requests', body),
	run:    (id: string, human_note?: string) =>
		api.post<TaskRequest>(`/task-requests/${id}/run`, { human_note }),
	reject: (id: string, human_note?: string) =>
		api.post<TaskRequest>(`/task-requests/${id}/reject`, { human_note }),
};
